### PR TITLE
[cmd/telemetrygen]: Allow float values for rate

### LINKED ISSUE #33982: Support float64 rates for telemetrygen
### Component(s)

cmd/telemetrygen

### Is your feature request related to a problem? Please describe.

Using float values in `--rate` returns the following error:

```
Error: invalid argument "0.1" for "--rate" flag: strconv.ParseInt: parsing "0.1": invalid syntax
```

The `--rate` flag for telemetrygen is an [Int64Var](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/562b7b66f6247eeeb0da8fd1a0e0c10dc5873212/cmd/telemetrygen/internal/common/config.go#L118), however this is ultimately [parsed to a `rate.Limit` value](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/562b7b66f6247eeeb0da8fd1a0e0c10dc5873212/cmd/telemetrygen/internal/metrics/metrics.go#L80), which is an alias for [float64](https://pkg.go.dev/golang.org/x/time/rate#Limit).

I propose changing this flag to a Float64Var to allow non-integer rates. My specific use case is to set this value to something like `0.2`, to send a metric once every 5 seconds. Currently, the slowest rate I can set is `1`, exporting a metric every second (which is an aggressive rate).

### Describe the solution you'd like

Support rates <1/second

### Describe alternatives you've considered

We could also use [`rate.Every()`](https://pkg.go.dev/golang.org/x/time/rate#Every) to specify something like this, which might be clearer. But this just parses to a `rate.Limit`, so I think it is just syntactical sugar

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by damemi**
**Description:**
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33982
Switch `--rate` flag in telemetrygen to support Float64 values

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33982

**Testing:** Added unit tests for float value cases

**Documentation:** N/A

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -278,7 +278,7 @@ docker-otelcontribcol:
 docker-telemetrygen:
 	GOOS=linux GOARCH=$(GOARCH) $(MAKE) telemetrygen
 	cp bin/telemetrygen_* cmd/telemetrygen/
-	cd cmd/telemetrygen && docker build --platform linux/$(GOARCH) -t telemetrygen:latest .
+	cd cmd/telemetrygen && docker build --platform linux/$(GOARCH) --build-arg="TARGETOS=$(GOOS)" --build-arg="TARGETARCH=$(GOARCH)" -t telemetrygen:latest .
```
Comment (atoulme) [inline Makefile:281]: Is this change related?

File: cmd/telemetrygen/internal/common/config.go [hunk 2]
```diff
@@ -115,7 +115,7 @@ func (c *Config) GetTelemetryAttributes() []attribute.KeyValue {
 // CommonFlags registers common config flags.
 func (c *Config) CommonFlags(fs *pflag.FlagSet) {
 	fs.IntVar(&c.WorkerCount, "workers", 1, "Number of workers (goroutines) to run")
-	fs.Int64Var(&c.Rate, "rate", 0, "Approximately how many metrics per second each worker should generate. Zero means no throttling.")
+	fs.Float64Var(&c.Rate, "rate", 0, "Approximately how many metrics per second each worker should generate. Zero means no throttling.")
```
Comment (atoulme) [inline cmd/telemetrygen/internal/common/config.go:118]:
```suggestion
	fs.Float64Var(&c.Rate, "rate", 0, "Approximately how many metrics/spans/logs per second each worker should generate. Zero means no throttling.")
```

File: Makefile (hunk 1 above)
Comment (damemi) [reply Makefile:281]: I can open this in a different PR, but I noticed it while building that `make docker-telemetrygen` is broken without it (at least for me):

```
$ make docker-telemetrygen
no Go files in //home/mikedame/go/src/github.com/open-telemetry/opentelemetry-collector-contrib
GOOS=linux GOARCH=amd64 make telemetrygen
make[1]: Entering directory '/home/mikedame/go/src/github.com/open-telemetry/opentelemetry-collector-contrib'
no Go files in /home/mikedame/go/src/github.com/open-telemetry/opentelemetry-collector-contrib
cd ./cmd/telemetrygen && GO111MODULE=on CGO_ENABLED=0 go build -trimpath -o ../../bin/telemetrygen_linux_amd64 \
	-tags "" .
make[1]: Leaving directory '/home/mikedame/go/src/github.com/open-telemetry/opentelemetry-collector-contrib'
cp bin/telemetrygen_* cmd/telemetrygen/
cd cmd/telemetrygen && docker build --platform linux/amd64 -t telemetrygen:latest .
Sending build context to Docker daemon   25.2MB
Step 1/10 : FROM alpine:latest as certs
 ---> 1d34ffeaf190
Step 2/10 : RUN apk --update add ca-certificates
 ---> Using cache
 ---> 958d3ffd0e84
Step 3/10 : FROM scratch
 ---> 
Step 4/10 : ARG USER_UID=10001
 ---> Using cache
 ---> 77b85f6e33ac
Step 5/10 : USER ${USER_UID}
 ---> Using cache
 ---> 22c6919286c2
Step 6/10 : ARG TARGETOS
 ---> Using cache
 ---> 742a0cf13877
Step 7/10 : ARG TARGETARCH
 ---> Using cache
 ---> 751f7303ea11
Step 8/10 : COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
 ---> Using cache
 ---> 3450c4e5656e
Step 9/10 : COPY ./telemetrygen_${TARGETOS}_${TARGETARCH} /telemetrygen
COPY failed: file not found in build context or excluded by .dockerignore: stat telemetrygen__: file does not exist
make: *** [Makefile:281: docker-telemetrygen] Error 1
```

Comment (atoulme) [reply Makefile:281]: If it's not related, it should go to a separate PR.

Comment (damemi) [reply Makefile]: @atoulme sure, moved it to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33989

**Comment by damemi** [comment]
@atoulme is this all set?

File: cmd/telemetrygen/telemetrygen_linux_amd64
[the remark this answers is not in the document]
Comment (damemi) [reply cmd/telemetrygen/telemetrygen_linux_amd64:1]: removed

File: .chloggen/telemetrygen-float-rate.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: telemetrygen
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Telemetrygen `--rate` flag changed from Int64  to Float64
```
[the remark this answers is not in the document]
Comment (damemi) [reply .chloggen/telemetrygen-float-rate.yaml]: done

File: cmd/telemetrygen/internal/traces/worker_test.go [hunk 4]
```diff
@@ -105,6 +105,36 @@ func TestRateOfSpans(t *testing.T) {
...
+	// sanity check
+	require.Len(t, syncer.spans, 0)
+
+	// test
+	require.NoError(t, Run(cfg, zap.NewNop()))
+
+	// verify
+	// the minimum acceptable number of spans for the rate of .3/sec for 3 seconds
+	assert.True(t, len(syncer.spans) >= 1, "there should have been more than 1 spans, had %d", len(syncer.spans))
+	// the minimum acceptable number of spans for the rate of .3/sec for 3 seconds
```
Comment (codeboten) [inline cmd/telemetrygen/internal/traces/worker_test.go:134]:
```suggestion
	// the maximum acceptable number of spans for the rate of .3/sec for 3 seconds
```

**Review by codeboten — APPROVED** [review]

File: cmd/telemetrygen/internal/logs/worker_test.go [hunk 5]
```diff
@@ -75,6 +75,28 @@ func TestRateOfLogs(t *testing.T) {
...
+	exp := &mockExporter{}
+
+	// test
+	require.NoError(t, Run(cfg, exp, zap.NewNop()))
+
+	// verify
+	// the minimum acceptable number of logs for the rate of .3/sec for 3 seconds
+	assert.True(t, len(exp.logs) >= 1, "there should have been 1 or more logs, had %d", len(exp.logs))
+	// the maximum acceptable number of logs for the rate of .3/sec for 3 seconds
+	assert.True(t, len(exp.logs) <= 2, "there should have been less than 3 logs, had %d", len(exp.logs))
```
Comment (jpkrohling) [inline cmd/telemetrygen/internal/logs/worker_test.go:97]: I would do this differently:

- Rate: 10/s
- Duration: 500ms
- Expected: more than 1, less than 7 (to give some room for failure)

This way, you don't need to run it for 3 seconds, while still being robust against slow runners. The way it is right now, we are introducing a 9s delay to every build we run (as opposed to 1.5s from the suggestion above).

Comment (damemi) [reply cmd/telemetrygen/internal/logs/worker_test.go:97]: I agree that adding 9s to the build isn't great (I wasn't too happy about that either), but the point of this is to test float values for the rate. In my specific use case, that's testing a rate <1, which I think is the only practical use case for float values anyway. Testing for output between 1-7 is a pretty wide range for the test, too.

So if the test is meant for values <1, then it will need at least 1s to get at least 1 span, am I understanding right? So would you be ok shrinking this to ~1s?

Comment (jpkrohling) [reply cmd/telemetrygen/internal/logs/worker_test.go:97]: True, can't rate be a float higher than 1? 5.5? We are still testing how many traces were generated based on a float value, so the float can be any float, no?

Comment (damemi) [reply cmd/telemetrygen/internal/logs/worker_test.go:97]: It can, but in practice there isn't much measurable difference between 5 and 6 spans/second, but there is a difference when you drop below 1 span/second -> 1 span/5 seconds. That's what I mean by the only practical use for float values, to invert the rate from spans/second to seconds/span.

With a rate >1 and a duration <1s the range of possible values is so wide (like you said, [1, 7]) that I don't see much value in the test. We don't need to test that just any float value works, because the pflag library does that for us. I'm trying to test a specific application of that float value.

In that case, I'm fine with just removing these tests from here and I can add them to our downstream code since it's my use case. I don't think there is much need for a new test besides that.

Comment (jpkrohling) [reply cmd/telemetrygen/internal/logs/worker_test.go:97]: To be honest, I'm not seeing good value for this cost: rate is already there, there's a test for it as well, and you are only changing the data type.

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 6]
```diff
@@ -91,13 +91,14 @@ func (w worker) simulateMetrics(res *resource.Resource, exporterFunc func() (sdk
...
 		}
 
-		if err := exporter.Export(context.Background(), &rm); err != nil {
-			w.logger.Fatal("exporter failed", zap.Error(err))
-		}
 		if err := limiter.Wait(context.Background()); err != nil {
 			w.logger.Fatal("limiter wait failed, retry", zap.Error(err))
 		}
 
+		if err := exporter.Export(context.Background(), &rm); err != nil {
```
Comment (jpkrohling) [inline cmd/telemetrygen/internal/metrics/worker.go:98]: I'm curious about this move: I don't see anything wrong with limiting it before exporting, but did you see a problem with exporting and then waiting?

Comment (damemi) [reply cmd/telemetrygen/internal/metrics/worker.go:98]: I think this was causing an off-by-one error in the amount of spans I was expecting. In the tests I was adding, I was expecting between 1 and 2 spans/logs/metric points (for the rate/duration of those tests), but kept getting 3. I believe firing off the first data point immediately was causing that extra point. Moving this seemed to fix it and it seemed more in line with the expectation to limit before sending.

File: cmd/telemetrygen/internal/logs/worker_test.go (hunk 5 above)
Comment (damemi) [reply cmd/telemetrygen/internal/logs/worker_test.go:97]: No problem, I agree, I removed these tests

**Review by jpkrohling — APPROVED** [review]

File: cmd/telemetrygen/internal/metrics/worker.go (hunk 6 above)
Comment (jpkrohling) [reply cmd/telemetrygen/internal/metrics/worker.go:98]: sounds good to me
